[WebIDL] Add WebIDL support for [Enumerated] extended attribute and related attributes [MissingValueDefault],[InvalidValueDefault]

rahimabdi opened the pull request:
#### 5c150f7a3b5880d65eb13e92222e2d177846693d
<pre>
[WebIDL] Add WebIDL support for [Enumerated] extended attribute and related attributes [MissingValueDefault],[InvalidValueDefault]
<a href="https://bugs.webkit.org/show_bug.cgi?id=295854">https://bugs.webkit.org/show_bug.cgi?id=295854</a>
<a href="https://rdar.apple.com/problem/156249490">rdar://problem/156249490</a>

Reviewed by Darin Adler.

To simplify .idl definitions of DOMString/DOMString? reflected enumerated attributes (<a href="https://html.spec.whatwg.org/#keywords-and-enumerated-attributes)">https://html.spec.whatwg.org/#keywords-and-enumerated-attributes)</a>,
and reduce manual generation of bindings code, this PR adds 3 new WebKitIDL extended attributes:
- [Enumerated]
- [MissingValueDefault]
- [InvalidValueDefault]

These attributes provide bindings codegen for C++ getters/setters using the specified enum. For example,
given a nullable enumerated attr that is limited to known values, its IDL definition would look like:

enum AttrEnum { &quot;value1&quot;, &quot;value2&quot;,...&quot;valueN&quot; }

[Reflect, Enumerated=AttrEnum] attribute DOMString? attr;

Optionally, a missing and/or invalid value default can be supplied to the enum definition, e.g.,:

[MissingValueDefault=&quot;value1&quot;, InvalidValueDefault=&quot;value2&quot;] enum AttrEnum { &quot;value1&quot;, &quot;value2&quot;...&quot;valueN&quot; }

In addition to adding TestObj.idl tests, the loading attribute for HTMLImageElement and HTMLIFrameElement now use these new extended attributes.

* Source/WebCore/bindings/scripts/IDLAttributes.json
* Source/WebCore/bindings/scripts/CodeGenerator.pm
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.h
* Source/WebCore/bindings/scripts/test/TestObj.idl
* Source/WebCore/html/HTMLImageElement.cpp
* Source/WebCore/html/HTMLImageElement.h
* Source/WebCore/html/HTMLImageElement.idl
* Source/WebCore/html/HTMLIFrameElement.cpp
* Source/WebCore/html/HTMLIFrameElement.h
* Source/WebCore/html/HTMLIFrameElement.idl

Canonical link: <a href="https://commits.webkit.org/298172@main">https://commits.webkit.org/298172@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/515458bf7c7f35d24df38a7ba4540679ac2c1639

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114504 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34250 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24713 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120668 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65224 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116393 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34879 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42810 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87023 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41931 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117452 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27790 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102831 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67414 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26975 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20958 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64350 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97167 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21072 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123875 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41517 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30981 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95841 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41894 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99022 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95625 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24365 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40790 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18628 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37553 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41396 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46906 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40982 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44289 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42732 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/bindings/scripts/IDLAttributes.json:
```diff
@@ -425,6 +425,22 @@
+            "contextsAllowed": ["attribute"],
+            "values": ["*"],
+            "notes" : "Optional invalid value default for enumerated attributes. Use via [ReflectInvalid=(\"value\")] for content attributes that specify both [Reflect] and [ReflectOnly]"
+            
```
```suggestion
```

weinig reviewed (commented):
I love the idea here of adding more that can be declaratively defined! Thanks for working on this.

I would like to reconsider the names though, and try to potentially align more closely with the HTML spec. The HTML spec calls these "enumerated attributes" (https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#enumerated-attribute). One possibility is lean into that, and actually use IDL enumerations for this. Taking one of the examples from the test:

```
{
...
    [Reflect=myattr, ReflectOnly=("Value1", "Value2", "Value3")] attribute DOMString reflectedStringLimitedToKnownValuesAttr;
...
}
```

It could look something like:

```
enum LimitedValues {
    "Value1", 
    "Value2", 
    "Value3"
}

{
...
    [Reflect=myattr, Enumerated=LimitedValues] attribute DOMString reflectedStringLimitedToKnownValuesAttr;
....
}
```

`enum` declarations in IDL are not actually exposed, so this would be fine from that perspective. 

You would still need missing and invalid though.

Not going to mark this as "request changes", as I don't think its necessary to take the direction I am suggesting.

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5888,6 +5954,7 @@ sub GenerateAttributeSetterBodyDefinition
 
             my $toNativeExpression = JSValueToNative($interface, $attribute, "value", $attribute->extendedAttributes->{Conditional}, "&lexicalGlobalObject", "lexicalGlobalObject", "thisObject", $globalObjectReference, $exceptionThrower);
             push(@$outputArray, "    auto nativeValueConversionResult = ${toNativeExpression};\n");
+
```
Let’s not add this blank line.
```suggestion
```

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
+            my @enumValues = ();
+
+            unless ($enumNameForAttribute) {
+                die "The {$attributeName} attribute does not have a valid enum specified. Ensure proper usage, e.g., [Reflect, Enumerated=myEnum...]\n";
```
Do we need the entire script to terminate when this happens? Maybe print the error and continue so we can see what other errors occur? Please take a look at how we handle other errors and try to keep it consistent.

rahimabdi replied:
Got it, changed this to an "print" statement (assuming `assert()` is not appropriate here).

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
+                die "The {$attributeName} attribute does not have a valid enum specified. Ensure proper usage, e.g., [Reflect, Enumerated=myEnum...]\n";
+            }
+
+            my $foundEnum = 0;
```
Do we need this? Can we use `@enumValues` for this purpose?

rahimabdi replied:
No, it's not needed. I've removed this `$foundEnum` check and simply go through all available `$interface->$enumerations`.

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
+
+            my $foundEnum = 0;
+            foreach my $enum (@$enumerations) {
+                next unless ref($enum) eq 'IDLEnum';
```
Why is this necessary? What other kinds of objects will be in the enumerations array?

rahimabdi replied:
Removed.

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
+                next unless ref($enum) eq 'IDLEnum';
+                my $enumName = $enum->{'IDLEnum::name'};
+                if ($enumName eq $enumNameForAttribute) {
+                    my $enumExtendedAttributes = $enum->{'IDLEnum::extendedAttributes'};
```
A single word name for this local variable is fine. The context is looping through enumerations and extracting the attributes so we don’t need to reiterate what kind of attributes these are in the local variable name.
```suggestion
                    my $attributes = $enum->{'IDLEnum::extendedAttributes'};
```

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
+                if ($enumName eq $enumNameForAttribute) {
+                    my $enumExtendedAttributes = $enum->{'IDLEnum::extendedAttributes'};
+                    $missingValueDefault = $enumExtendedAttributes->{'MissingValueDefault'};
+                    $missingValueDefault = undef if !defined($missingValueDefault) || $missingValueDefault eq '"VALUE_IS_MISSING"';
```
If it’s not defined we don’t need to make it not defined again. Maybe we can use this simpler idiom?
```suggestion
                    undef $missingValueDefault if $missingValueDefault eq '"VALUE_IS_MISSING"';
```
It’s also really surprising that the value is `"VALUE_IS_MISSING"` with the quote marks. Do you know why the parser works that way? I also don’t see a test case covering this. Do we need to support it or can we just make that a parsing error?

rahimabdi replied:
These should definitely not have quotes, fixed.

In case you were asking why there is generally a "VALUE_IS_MISSING", it has a couple purposes from my understanding:

- When parsing extended attributes that use `[Reflect]` and related extended attributes, it serves as a placeholder value for the reflected attribute name if it's missing from the attribute declaration on the same line. For example:

`[CEReactions=Needed, Reflect=aria_busy] attribute DOMString? ariaBusy;    // parser picks up aria_busy for $attribute->extendedAttributes->{Reflect} which eventually becomes aria_busyAttr`

`[Reflect] attribute DOMString? role;    // $attribute->extendedAttributes->{Reflect} is "VALUE_IS_MISSING" at first and a later check ensures it falls back to $attribute->name which is role`

- It's used for several extended attributes on interfaces to check whether they have been supplied with a value, e.g.,:

`assert("Must specify value for EnabledForWorld.") if $context->extendedAttributes->{CustomEnabledBy} eq "VALUE_IS_MISSING";`

`assert("Must specify value for EnabledForWorld.") if $context->extendedAttributes->{EnabledForWorld} eq "VALUE_IS_MISSING";`

`assert("Must specify value for EnabledBySetting.") if $context->extendedAttributes->{EnabledBySetting} eq "VALUE_IS_MISSING";`

I don't believe we need test cases (since it's current parser behavior) but if you suggest improving this, I can raise a separate bug.

rahimabdi replied:
The Perl interpreter seems to complain with this ("Use of uninitialized value $invalidValueDefault in string eq"):

`undef $invalidValueDefault if $invalidValueDefault eq '"VALUE_IS_MISSING"';`

I changed it to this for a valid string comparison:

`undef $invalidValueDefault if ($invalidValueDefault // '') eq '"VALUE_IS_MISSING"';`

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
 
+        if ($attribute->extendedAttributes->{Reflect} && $attribute->extendedAttributes->{Enumerated}) {
+            my $attributeName = $attribute->extendedAttributes->{Reflect};
+            $attributeName = $attribute->name if !$attributeName || $attributeName eq "VALUE_IS_MISSING";
```
This `VALUE_IS_MISSING` special value name is frustrating. It’s resulting in extra code. Can we improve how that works to avoid this strange extra code?

Also, I don’t see a test case covering this. Need to have that.

rahimabdi replied:
Would you recommend a separate bug/PR or should I fix it here?

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
+                push(@$outputArray, "        returnValue = AtomString(\"$invalidValueDefault\"_s);\n");
+            } else {
+                push(@$outputArray, "        returnValue = nullAtom();\n");
+            }
```
```suggestion
            if (defined $invalidValueDefault) {
                push @$outputArray, "    else\n";
                push @$outputArray, "        returnValue = AtomString(\"$invalidValueDefault\"_s);\n";
            }
```

rahimabdi replied:
Updated logic to use enum utils such as `parseEnumerationFromString()`, no longer relevant.

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
+                die "The enum {$enumNameForAttribute} was not found in the defined enums.\n";
+            }
+
+            push(@$outputArray, "    const AtomString& contentAttributeValue = impl.attributeWithoutSynchronization(WebCore::HTMLNames::${attributeName}Attr);\n");
```
This code has a mix of styles for the push calls. I suggest sticking with one consistent style. Maybe the one without parentheses.
```suggestion
            push @$outputArray, "    const AtomString& contentAttributeValue = impl.attributeWithoutSynchronization(WebCore::HTMLNames::${attributeName}Attr);\n";
```

rahimabdi replied:
Updated to `push()` with parenthesis as that is most prevalent (there's only 6 instances of `push` without parenthesis; although both are allowed, I can fix that in this PR as well).

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
+
+            my $firstEnumValue = shift @enumValues;
+            push @$outputArray, "    else if (equalLettersIgnoringASCIICase(contentAttributeValue, \"$firstEnumValue\"_s))\n";
+            push @$outputArray, "        returnValue = AtomString(\"$firstEnumValue\"_s);\n";
```
This isn’t needed. We’re not handling the first value different than the rest.
```suggestion
            my $firstEnumValue = shift @enumValues;
            push @$outputArray, "    else if (equalLettersIgnoringASCIICase(contentAttributeValue, \"$firstEnumValue\"_s))\n";
            push @$outputArray, "        returnValue = AtomString(\"$firstEnumValue\"_s);\n";
```

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
+                push(@$outputArray, "        returnValue = AtomString(\"$missingValueDefault\"_s);\n");
+            } else {
+                push(@$outputArray, "        returnValue = nullAtom();\n");
+            }
```
A newly defined `AtomString` is already null so no need to assign it that value. It would be best to not generate any code at all, not even a check for null, but it’s probably tricky to do that given how we emit an `else` below and how the empty check relies on us already having checked for null, so I guess I’d suggest this.
```suggestion
            push(@$outputArray, "    if (contentAttributeValue.isNull())\n");
            if (defined $missingValueDefault) {
                push(@$outputArray, "        returnValue = AtomString(\"$missingValueDefault\"_s);\n");
            } else {
                push(@$outputArray, "        ;\n");
            }
```

darinadler commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5670,6 +5670,72 @@ sub GenerateAttributeGetterBodyDefinition
+            }
+
+            push(@$outputArray, "    const AtomString& contentAttributeValue = impl.attributeWithoutSynchronization(WebCore::HTMLNames::${attributeName}Attr);\n");
+            push(@$outputArray, "    AtomString returnValue;\n");
```
Other generated code uses `result` for this, not `returnValue`, so I suggest we do the same:
```suggestion
            push(@$outputArray, "    AtomString result;\n");
```

darinadler commented on Source/WebCore/bindings/scripts/IDLParser.pm:
```diff
@@ -2401,6 +2401,12 @@ sub parseExtendedAttributeRest2
+    if ($next->type() == StringToken) {
+        my $token = $self->getToken();
+        my $value = $token->value();
+        $value =~ s/^"(.*)"$/$1/;
```
Why are the quotation marks around the string optional? I don’t see tests for with and without them.

rahimabdi replied:
I've added tests to allow for the enum missing/invalid value default to support them, i.e., these are both valid:

[MissingValueDefault="someVal"] myEnum

[MissingValueDefault=someVal] my Enum

darinadler commented on Source/WebCore/bindings/scripts/IDLAttributes.json:
```diff
@@ -173,6 +173,11 @@
+        "EmptyValueDefault": {
+            "contextsAllowed": ["enum"],
+            "values": ["*"],
+            "notes" :"optional value default for enumerated content attributes when the attribute is present but without a value. Must be used only for content attributes that specify both [Reflect, Enumerated]"
```
This description is a bit wordy and repetitive. I don’t see notes on other extended attributes in this file. Is it really helpful to have this note? Can we instead refer to some documentation? I’m not sure this file is a good place to store documentation for these extended attributes.

I’m not thrilled when the note repeats what the text does. For example, the fact that it’s in an enumeration is already said above by the `contextsAllowed`.

Here we have lowercase letter for the start of the sentence. Also we have a period on the first sentence but not the second.

This comment talks about content attributes, but this is on an enumeration, not on the attribute, so the comment doesn’t really make total sense. You can put this on any enumeration, and it affects what happens on certain content attributes, so the comment and documentation should talk about it that way, instead of talking about this as if it was an extended attribute for a content attribute.

If we keep this, please fix the colon:
```suggestion
          "notes": "optional value default for enumerated content attributes when the attribute is present but without a value. Must be used only for content attributes that specify both [Reflect, Enumerated]"
```

rahimabdi replied:
Agreed, I've removed these notes and added a link to the enumerated attributes section of HMTL spec.

darinadler commented on Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
```diff
@@ -4261,6 +4316,88 @@ JSC_DEFINE_CUSTOM_SETTER(setJSTestObj_reflectedStringAttr, (JSGlobalObject* lexi
+    else if (equalLettersIgnoringASCIICase(contentAttributeValue, "value2"_s))
+        returnValue = AtomString("value2"_s);
+    else if (equalLettersIgnoringASCIICase(contentAttributeValue, "value3"_s))
+        returnValue = AtomString("value3"_s);
```
This is a less efficient way to do the same thing that `parseEnumerationFromString` does with `SortedArrayMap`. Would be better to share that logic.

rahimabdi replied:
Much better approach, thanks! Updated the codegen to use `parseEnumerationFromString()` and `convertEnumerationToString()` utils.

darinadler commented on Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
```diff
@@ -4261,6 +4316,88 @@ JSC_DEFINE_CUSTOM_SETTER(setJSTestObj_reflectedStringAttr, (JSGlobalObject* lexi
+    else if (equalLettersIgnoringASCIICase(contentAttributeValue, "value3"_s))
+        returnValue = AtomString("value3"_s);
+    else if (equalLettersIgnoringASCIICase(contentAttributeValue, ""_s))
+        returnValue = AtomString(""_s);
```
This is dead code. The is the same case as `contentAttributeValue.isEmpty()` above. We should figure out whether we want to allow this to happen or not.

darinadler commented on Source/WebCore/bindings/scripts/test/TestObj.idl:
```diff
@@ -46,6 +46,8 @@ enum TestObjKind { "quick", "dead" };
 enum ObjSize { "small", "much-much-larger" };
 enum TestConfidence { "high", "kinda-low" };
 
+[MissingValueDefault="value1", InvalidValueDefault="value2", EmptyValueDefault="value3"] enum attributeValues { "value1", "value2", "value3", "" };
```
Doesn’t really make logical sense to have `EmptyValueDefault` and also an empty value.

We need test cases for more different combinations, instead of just a single case where all the optional defaults are used.

rahimabdi replied:
I added these examples as ARIA will have attributes that need to support the empty string as an explicit keyword but agree that having both doesn't make sense. I've removed `[EmptyValueDefault]` support from this PR per your suggestion to make smaller changes.

darinadler reviewed (commented):
Some comments about the draft version.

In case this wasn’t obvious from our earlier requests please also include one actual use of any feature you’re adding for something in the DOM, don’t just add all the features with tests.

I’d prefer a patch that adds a bit less and uses what’s added. Then we can add more as we use more.

darinadler commented on Source/WebCore/html/HTMLIFrameElement.idl:
```diff
@@ -49,3 +49,6 @@
 };
+
+[MissingValueDefault="eager", InvalidValueDefault="eager"] enum HTMLIFrameElementLoadingValues { "lazy", "eager" };
+
```
```suggestion
```

darinadler commented on Source/WebCore/html/HTMLImageElement.idl:
```diff
@@ -63,3 +63,5 @@
 };
 
+[MissingValueDefault="eager", InvalidValueDefault="eager"] enum HTMLImageElementLoadingValues { "lazy", "eager" };
+
```
```suggestion
```

darinadler reviewed (approved):
Looks fine. The old code style might be slightly more efficient with the way it creates the atom strings.

rahimabdi committed:
[WebIDL] Add WebIDL support for [Enumerated] extended attribute and related attributes [MissingValueDefault],[InvalidValueDefault]
https://bugs.webkit.org/show_bug.cgi?id=295854
rdar://problem/156249490

Reviewed by Darin Adler.

To simplify .idl definitions of DOMString/DOMString? reflected enumerated attributes (https://html.spec.whatwg.org/#keywords-and-enumerated-attributes),
and reduce manual generation of bindings code, this PR adds 3 new WebKitIDL extended attributes:
- [Enumerated]
- [MissingValueDefault]
- [InvalidValueDefault]

These attributes provide bindings codegen for C++ getters/setters using the specified enum. For example,
given a nullable enumerated attr that is limited to known values, its IDL definition would look like:

enum AttrEnum { "value1", "value2",..."valueN" }

[Reflect, Enumerated=AttrEnum] attribute DOMString? attr;

Optionally, a missing and/or invalid value default can be supplied to the enum definition, e.g.,:

[MissingValueDefault="value1", InvalidValueDefault="value2"] enum AttrEnum { "value1", "value2"..."valueN" }

In addition to adding TestObj.idl tests, the loading attribute for HTMLImageElement and HTMLIFrameElement now use these new extended attributes.

* Source/WebCore/bindings/scripts/IDLAttributes.json
* Source/WebCore/bindings/scripts/CodeGenerator.pm
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.h
* Source/WebCore/bindings/scripts/test/TestObj.idl
* Source/WebCore/html/HTMLImageElement.cpp
* Source/WebCore/html/HTMLImageElement.h
* Source/WebCore/html/HTMLImageElement.idl
* Source/WebCore/html/HTMLIFrameElement.cpp
* Source/WebCore/html/HTMLIFrameElement.h
* Source/WebCore/html/HTMLIFrameElement.idl

Canonical link: https://commits.webkit.org/298172@main
